makepatch.man
.TH MAKEPATCH 1 "93/09/11" "Version 1.8"
.SH NAME
makepatch \- create patch diffs between two versions of source
.SH SYNOPSIS
.B makepatch
.RI [ options ]
.I old new
.PP
.B makepatch
\-filelist
.RI [ options ] 
.I manifest

.SH DESCRIPTION
.I Makepatch
generates a set of differences between two files or two sets of files
maintained in two different directories and prints the results to
\fIstdout\fP.
This resulting output is suitable for use by the 
.IR patch (1)
program to update copies of the target file(s) from the \fIold\fP
to the \fInew\fP version.
.LP
Features of this utility include:
.na
.TP 3n
\- 
Recursive descend through sub-directories.
.TP 3n
\- 
Generation of commands to remove obsolete files.
.TP 3n
\- 
Automatic handling of the \fIpatchlevel.h\fP file first.
.TP 3n
\- 
Automatic inclusion of \fIIndex:\fP and \fIPrereq:\fP lines.
.TP 3n
\- 
Ability to utilize specified \fImanifest\fP file(s).
.ad

.SH ARGUMENTS
.TP 6
.I old
This is the name of either a single file or else a directory which
contains copies of the older version of the target file(s); in
other words, copies of the file(s) \fIprior\fP to any modifications.
.TP 
.I new
This is the name of either a single file or else a directory which
contains copies of the newer version of the target file(s); in
other words, copies of the file(s) \fIafter\fP the modifications have
been made.
A 
.IR rm (1)
command will automatically be generated for every
\fIold\fP file that no longer has a corresponding \fInew\fP version.

.SH "MAKEPATCH OPTIONS"
.TP 6
\fB\-diff\fP \fIcmd\fP
If specified, \fIcmd\fP is the command to be used to generate the
differences between the two versions of the files.
If not specified, this command defaults to "\fBdiff \-c\fP".
.TP 6
\fB\-patchlevel\fP \fIpfile\fP
If specified, \fIpfile\fP indicates an alternate file that is to be
used in lieu of "\fBpatchlevel.h\fP".
.TP 6
\fB\-man\fP[\fBifest\fP] \fImfile\fP
If specified, \fImfile\fP indicates the name of the manifest file
which consists of a list of the files contained in both the \fIold\fP
and the \fInew\fP directories.
.TP 6
\fB\-oldman\fP[\fBifest\fP] \fIomfile\fP
If specified, \fIomfile\fP indicates the name of the manifest file
which consists of a list of the files contained in the \fIold\fP
directory.
This option is designed to be used in conjunction with the
\%\fB-newmanifest\fP option.
Note that the \fIold\fP and \fInew\fP directories must still be
indicated.
.TP 6
\fB\-newman\fP[\fBifest\fP] \fInmfile\fP
If specified, \fInmfile\fP indicates the name of the manifest file
which consists of a list of the files contained in the \fInew\fP
directory.
This option is designed to be used in conjunction with the
\%\fB-oldmanifest\fP option.
Note that the \fIold\fP and \fInew\fP directories must still be
indicated.
.TP 6
\fB\-follow\fP
If specified, symbolic links to directories are traversed as if they
were real directories.

.SH "FILELIST OPTIONS"
.TP 6
.BR \- [ file ] list 
This option instructs 
.I makepatch
to read a manifest file, and output the list of files included in
this manifest. This option is useful to turn the contents of a
manifest file into a list of files suitable for other programs.
.TP 6
\fB\-man\fP[\fBifest\fP] \fImfile\fP
If specified, \fImfile\fP indicates the name of the manifest file to
be used. Alternatively, the name of the manifest file may follow the
command line options.
.TP 6
.B \-prefix 
.I string
Every entry in the manifest file is prefixed with
.I string
before it is written to 
.IR stdout .
.TP 6
.B \-nosort
Retain the order of filenames from the manifest file.

.SH "GENERAL OPTIONS"
.TP 6
.B \-verbose
This is the default mode which displays information
concerning \fBmakepatch\fP's activity to \fIstderr\fP.
.TP 6
.B \-quiet
The opposite of \fB-verbose\fP.
This instructs \fImakepatch\fP to suppress the display of
activity information.
.TP 6
.B \-help
This causes a short help message to be displayed, after which the
program immediately exits.

.SH "MANIFEST FILES"
Although there is no formal standard for manifest files, the following
rules apply:
.TP 3n
\-
If the second line from the manifest file looks like a separator
line (e.g. it is empty, or contains only dashes), it is discarded and
so is the first line. 
.TP 3n
\-
Empty lines and lines that start with a 
.B #
are ignored.
.TP 3n
\-
If there are multiple space-separated ``words'' on a line, the
first word is considered to be the filename.

.SH EXAMPLES
Suppose you have a directory tree 
.B emacs\-18.58
containing the sources for GNU Emacs 18.58, and a directory tree
.B emacs\-18.59
containing the sources for GNU Emacs 18.59. The following command will
generate the patch file needed to transform the 18.58 sources into
18.59: 

.in +3n
.na
makepatch emacs\-18.58 emacs\-18.59 > emacs\-18.58\-18.59.diff
.in
.ad

This is one way to generate and use manifest files:

.in +3n
.na
(cd emacs\-18.58; find . \-type f \-print > MANIFEST)
.br
(cd emacs\-18.59; find . \-type f \-print > MANIFEST)
.br
makepatch \e
.in +3n
\-oldmanifest emacs\-18.58/MANIFEST \e
.br
\-newmanifest emacs\-18.59/MANIFEST \e
.br
emacs\-18.58 emacs\-18.59 > emacs\-18.58\-18.59.diff
.in -6n
.ad

The following example transforms the manifest file into a list of
files suitable for GNU tar. Note the trailing
.B /
in the prefix string:

.na
.in +3n
makepatch \-filelist \-prefix emacs\-18.59/ emacs\-18.59/MANIFEST |
.in +3n
gtar \-Zcvf emacs\-18.59.tar.Z \-T \-Op
.in -6n
.ad

.SH "SEE ALSO"
.IR diff (1),
.IR patch (1),
.IR  perl (1),
.IR rm (1).

.SH AUTHORS
Johan Vromans (jv@mh.nl) wrote the program.
.br
Jeffery Small (jeff@cjsa.uucp) donated the base version of this manual
page that inspired me to complete it.


CHANGES
Overall changes in makepatch version 2.0
----------------------------------------

 * Generates a perl program that applies the patches by running the
   'patch' program. Optionally, a shell script can be generated as with
   the 1.x version.
   The generated perl program requires a normal perl5 installation and
   the 'patch' program. No other tools (like shells) are required.

   The generated perl program is the concatenation of:
    - a perl 'Patcher' program;
    - a shell preamble;
    - the actual 'patch' input.

   If running through perl fails due to perl contraints, the perl
   preamble can be stripped off and the remainder used as a shell
   script. If that fails due to shell constraints, new
   files/directories should be created manually after which the shell
   script can be fed to the 'patch' program.

 * To avoid most common problems with buggy versions of the 'patch'
   program, the patch data is relocated to the current directory. 
   This obsoletes the command line options -fixpath and -fixallpath.

   As a consequence, when a patch is generated from two directories,
   e.g.:

     makepatch src-1.0 src-1.1 > the-patch

   it must be applied *IN* the directory src-1.0, e.g.

     cd src-1.0; perl the-patch

   When fed to patch directly, no -p option must be used:

     cd src-1.0; patch < the-patch

 * The generated perl program will apply some heuristics to determine
   if it is in the correct directory before attempting to apply the
   patch.

 * The generated perl program will verify that the patch is complete,
   i.e. not truncated or currupted.

 * Program options can be specified in an environment variable, a file
   as well as on the command line.

 * Makepatch can take a standard source archive (.tar.gz, .tar.bz2,
   .tar or .zip) as input instead of the sources directories.

 * MANIFEST files are automatically used unless suppressed with the
   new -nomanifest option.

Changes in makepatch version 1.96 (2.0BETA)
-------------------------------------------

 * Extracted archives do not need to produce a single root directory.

 * Use 'Getopt::Long' format options file.

 * Add 'bzip2' style compression.

 * Hide filelist function.

Changes in makepatch version 1.95 (2.0BETA)
-------------------------------------------
This is makepatch 1.94.

Changes in makepatch version 1.94 (2.0BETA)
-------------------------------------------

 * Use temp dir instead of temp files. Clean up after use.

 * Add option -automanifest to automatically use $opt_manifest files if
   present.

 * Add option -nomanifest to suppress using manifest files.

 * Relocate the patch to current dir. Eliminates old -fixpath and
   -fixallpath options. Change 'patch -p1' to 'patch -p0'.

 * Include (most) option settings in the output.

 * Allow most option flags to be negatable.

 * Use smail style settings in the options file.

 * Check for end of patch, corruption and truncation before
   applying the patch.

 * Unpack .tar .tar.gz .tgz .zip archives automatically.

Changes in makepatch version 1.93 (2.0BETA)
-------------------------------------------

 * Add env.var. MAKEPATCHINIT.
   Add reading .makepatchrc and --rcfile option.
   Add perl generation.

Changes in makepatch version 1.92 (2.0BETA)
-------------------------------------------

 * Add -[no]recurse option.

Changes in makepatch version 1.91 (2.0BETA)
-------------------------------------------

 * Allow -exclude to handle [..] style wildcards.
   Check for dangerous filenames (with '"$`).

Changes in makepatch version 1.90 (2.0BETA)
-------------------------------------------

 * Rework for Perl 5.

 * Add -exclude, -exclude_regex, -exclude-vc.


makepatch.pl
#!/usr/local/bin/perl
# makepatch.pl -- generate batch of patches.
$RCS_Id = '$Id: makepatch.pl,v 1.10 1995/10/29 11:14:46 jv Exp $ ';
# Author          : Johan Vromans
# Created On      : Tue Jul  7 20:39:39 1992
# Last Modified By: Johan Vromans
# Last Modified On: Sun Oct 29 12:13:52 1995
# Update Count    : 136
# Status          : Released to USEnet.
#
# Generate a patch from two files or directories.
#
# Resembles "diff -c -r -N", but:
#
#   - always recursive
#   - handles 'patchlevel.h' first
#   - supplies 'Index:' and 'Prereq:' lines
#   - can use manifest file
#   - generates shell commands to remove files
#   - manipulates manifest files
#
################################################################
#
# Usage:
# 
#   makepatch <old-dir> <new-dir
# 
#     This will compare all files in <new-dir> against the files in
#     <old-dir>, and generate a bunch of patches to transform every
#     file in <old-dir> into the corresponding one in <new-dir>.
#     Files that appear in <new-dir> but not in <old-dir> are created.
#     For files that appear in <old-dir> but not in <new-dir>
#     'rm'-commands are generated at the beginning of the patch.
# 
# Using MANIFEST files:
# 
#   makepatch -oldmanifest <oldmanifest> -newmanifest <newmanifest> \
#           <new-dir> <old-dir>
# 
#     <oldmanifest> and <newmanifest> list the files in <old-dir>
#     and <new-dir> that are to be examined.
#     Only the files that are named will be examined. 
#     <oldmanifest> should contain the names of the files relative to
#     <old-dir> and <newmanifest> should contain the names of the files
#     relative to <new-dir>.
# 
#   makepatch -manifest <manifest> <new-dir> <old-dir>
# 
#     This is a simplified form of the above example.
#     <manifest> applies to both <old-dir> and <new-dir>.
# 
#   makepatch -filelist [ -prefix xxx ] manifest
#
#     The filenames are extracted from the manifest file,
#     optionally prefixed, sorted and written to standard output.
#
# Examples:
# 
#   % makepatch -verbose emacs-18.58 emacs-18.59 > emacs-18.58-18.59.diff
# 
#   % (cd emacs-18.58; find . -type f -print > MANIFEST)
#   % (cd emacs-18.59; find . -type f -print > MANIFEST)
#   % makepatch -verbose \
#         -oldmanifest emacs-18.58/MANIFEST \
#         -newmanifest emacs-18.59/MANIFEST \
#         emacs-18.58 emacs-18.59 > emacs-18.58-18.59.diff
#
#   % makepatch -filelist -prefix emacs-18.59/ emacs-18.59/MANIFEST |
#	gtar -zcvf emacs-18.59.tar.Z -T -
#
################################################################

################ Common stuff ################

# $LIBDIR = $ENV{'LIBDIR'} || '/usr/local/lib/sample';
# unshift (@INC, $LIBDIR);
# require 'common.pl';
$my_package = 'Sciurix';
($my_name, $my_version) = $RCS_Id =~ /: (.+).pl,v ([\d.]+)/;
$my_version .= '*' if length('$Locker:  $ ') > 12;

################ Program parameters ################

&stdio;
&options;
($old, $new) = @ARGV;

print STDERR ("This is $my_name version $my_version\n") if $opt_verbose;

if ( defined $opt_filelist ) {
    @new = &domanifest (shift (@ARGV));
    foreach ( @new ) {
	print STDOUT ($opt_prefix, $_, "\n");
    }
    exit (0);
}

$tmpfile = $ENV{"TMPDIR"} || "/usr/tmp";
$thepatch = "$tmpfile/mp$$.p";
$tmpfile .= "/mp$$.t";
open (PATCH, ">$thepatch") || die ("$thepatch: $!\n");
$patched = $created = 0;
&doit ($old, $new);
&wrapup;
exit (0);

################ Subroutines ################

sub doit {
    local ($old, $new) = @_;

    if ( -f $old && -f $new ) {
	# Two files.
	if ( $opt_verbose ) {
	    print STDERR ("Old file = $old.\n",
			  "New file = $new.\n");
	}
	&dodiff ("", $old, "", $new);
    }
    elsif ( -f $old && -d $new ) {
	# File and dir -> File and dir/File.
	$new = ( $new =~ m|^\./?$| ) ? "" : "$new/";
	if ( $opt_verbose ) {
	    print STDERR ("Old file = $old.\n",
			  "New file = $new$old.\n");
	}
	&dodiff ("", $old, $new, $old);
    }
    elsif ( -f $new && -d $old ) {
	$old = ( $old =~ m|^\./?$| ) ? "" : "$old/";
	if ( $opt_verbose ) {
	    print STDERR ("Old file = $old$new.\n",
			  "New file = $new.\n");
	}
	&dodiff ($old, $new, "", $new);
    }
    else {
	# Should be two directories.
	local (@old, @new);
	if ( defined $opt_oldmanifest ) {
	    @old = &domanifest ($opt_oldmanifest);
	}
	else {
	    @old = &make_filelist ($old);
	}
	if ( defined $opt_newmanifest ) {
	    @new = &domanifest ($opt_newmanifest);
	}
	else {
	    @new = &make_filelist ($new);
	}

	$new = ( $new =~ m|^\./?$| ) ? "" : "$new/";
	$old = ( $old =~ m|^\./?$| ) ? "" : "$old/";

	if ( $opt_verbose ) {
	    local ($old) = $old; chop ($old);
	    local ($new) = $new; chop ($new);
	    print STDERR ("Old dir = $old, file list = ",
			  defined $opt_oldmanifest ? $opt_oldmanifest : "<*>",
			  ", ", 0+@old, " files.\n");
	    print STDERR ("New dir = $new, file list = ",
			  defined $opt_newmanifest ? $opt_newmanifest : "<*>",
			  ", ", 0+@new, " files.\n");
	}
	if ( $opt_debug ) {
	    print STDERR ("Old: @old\nNew: @new\n");
	}

	# Handle patchlevel file first.
	$opt_patchlevel = (grep (/patchlevel\.h/, @new))[0]
	    unless defined $opt_patchlevel;

	if ( defined $opt_patchlevel && $opt_patchlevel ne "" ) {
	    if ( ! -f "$new$opt_patchlevel" ) {
		die ("$new$opt_patchlevel: $!\n");
	    }
	    if ( -f "$old$opt_patchlevel" ) {
		&dodiff ($old, $opt_patchlevel, $new, $opt_patchlevel);
	    }
	    else {
		$created++;
		&dodiff ("", "/dev/null", $new, $opt_patchlevel);
	    }
	}
	else {
	    undef $opt_patchlevel;
	}

	# Process the filelists.
	while ( @old + @new ) {

	    $o = shift (@old) unless defined $o;
	    $n = shift (@new) unless defined $n;
	    
	    if ( defined $n && (!defined $o || $o gt $n) ) {
		# New file.
		if ( defined $opt_patchlevel && $n eq $opt_patchlevel ) {
		    undef $opt_patchlevel;
		}
		else {
		    $created++;
		    &dodiff ("", "/dev/null", $new, $n);
		}
		undef $n;
	    }
	    elsif ( !defined $n || $o lt $n ) {
		# Obsolete (removed) file.
		push (@goners, $o);
		undef $o;
	    }
	    elsif ( $o eq $n ) {
		# Same file.
		if ( defined $opt_patchlevel && $n eq $opt_patchlevel ) {
		    undef $opt_patchlevel;
		}
		else {
		    &dodiff ($old, $o, $new, $n);
		}
		undef $n;
		undef $o;
	    }
	}
    }
}

sub make_filelist {
    local ($dir, $disp) = @_;

    # Return a list of files, sorted, for this directory.
    # Recurses.

    local (@ret);
    local (*DIR);
    local (@tmp);
    local ($fname);

    $disp = "" unless defined $disp;

    print STDERR ("+ recurse $dir\n") if $opt_trace;
    opendir (DIR, $dir) || die ("$dir: $!\n");
    @tmp = sort (readdir (DIR));
    closedir (DIR);
    print STDERR ("Dir $dir: ", 0+@tmp, " entries\n") if $opt_debug;

    @ret = ();
    foreach $file ( @tmp ) {

	# Skip unwanted files.
	next if $file =~ /^\.\.?$/; # dot and dotdot
	next if $file =~ /~$/;	# editor backup files

	# Push on the list.
	$fname = "$dir/$file";
	if ( -d $fname && ( $opt_follow || ! -l $fname ) ) {
	    # Recurse.
	    push (@ret, &make_filelist ($fname, "$disp$file/"));
	}
	elsif ( -f _ ) {
	    push (@ret, $disp . $file);
	}
	else {
	    print STDERR ("Ignored $fname: not a file\n");
	}
    }
    @ret;
}

sub domanifest {
    local ($man) = @_;
    local (*MAN);
    local (@ret) = ();

    open (MAN, $man) || die ("$man: $!\n");
    while ( <MAN> ) {
	if ( $. == 2 && /^[-=_\s]*$/ ) {
	    @ret = ();
	    next;
	}
	next if /^#/;
	next unless /\S/;
	$_ = $` if /\s/;
	push (@ret, $_);
    }
    close (MAN);
    @ret = sort @ret unless defined $opt_nosort;
    @ret;
}

sub dodiff {
    local ($olddir, $old, $newdir, $new) = @_;

    # Produce a patch hunk.

    local ($cmd) = "$opt_diff '$olddir$old' '$newdir$new'";
    print STDERR ("+ ", $cmd, "\n") if $opt_trace;
    $result = system ("$cmd > $tmpfile");
    printf STDERR ("+> result = 0x%x\n", $result) 
	if $result && $opt_debug;

    if ( $result && $result < 128 ) {
	&wrapup (($result == 2 || $result == 3) 
		 ? "User request" : "System error");
	exit (1);
    }
    return unless $result == 0x100;	# no diffs
    $patched++;

    # print PATCH ($cmd, "\n");
    print PATCH ("Index: ", $new, "\n");

    # Try to find a prereq.
    # The RCS code is based on a suggestion by jima@netcom.com, who also
    # pointed out that patch requires blanks around the prereq string.
    open (OLD, $olddir . $old);
    while ( <OLD> ) {
	next unless (/\@\(\#\)/		# SCCS header
		     || /\$Header:/ 	# RCS Header
		     || /\$Id:/); 	# RCS Header
	next unless $' =~ /\s\d+(\.\d+)*\s/; # e.g. 5.4
	print PATCH ("Prereq: $&\n");
	last;
    }
    close (OLD);

    # Copy patch.
    open (TMP, $tmpfile);
    print PATCH <TMP>;
    close (TMP);
}

sub wrapup {
    local ($reason) = @_;

    if ( defined $reason ) {
	print STDERR ("*** Aborted: $reason ***\n");
    }
    if ( $opt_verbose ) {
	local ($goners) = scalar (@goners);
	print STDERR ("Collecting: $patched patch",
		      $patched == 1 ? "" : "es");
	print STDERR (" ($created new file", 
		      $created == 1 ? "" : "s", ")") if $created;
	print STDERR (", $goners goner", 
		      $goners == 1 ? "" : "s") if $goners;
	print STDERR (".\n");
    }
    if ( @goners ) {
	print STDOUT 
	    ("# Please remove the following file",
	     @goners == 1 ? "" : "s", " before applying this patch.\n",
	     "# (You can feed this patch to 'sh' to do so.)\n",
	     "\n");
	foreach ( @goners ) {
	    print STDOUT ("rm -f ", $_, "\n");
	}
	print STDOUT ("exit\n\n");
    }

    # Copy patch.
    open (PATCH, $thepatch);
    print while <PATCH>;
    close (PATCH);

    # Cleanup.
    unlink ($tmpfile, $thepatch);
}

sub stdio {
    # Since output to STDERR seems to be character based (and slow),
    # we connect STDERR to STDOUT if they both point to the terminal.
    if ( -t STDOUT && -t STDERR ) {
	close (STDERR);
	open (STDERR, '>&STDOUT');
	select (STDERR); $| = 1;
	select (STDOUT);
    }
}

sub options {
    local ($opt_manifest);
    local ($opt_quiet);

    # Defaults...
    $opt_diff = "diff -c";
    $opt_verbose = 1;
    $opt_follow = 0;

    # Process options, if any...
    if ( $ARGV[0] =~ /^-/ ) {
	require "newgetopt.pl";

	# Aliases.
	*opt_man = *opt_manifest;
	*opt_oldman = *opt_oldmanifest;
	*opt_newman = *opt_newmanifest;
	*opt_v = *opt_verbose;
	*opt_list = *opt_filelist;

	if ( ! &NGetOpt ("patchlevel=s", "diff=s", 
			 "manifest=s", "newmanifest=s", "oldmanifest=s",
			 "man=s", "newman=s", "oldman=s", "follow",
			 "list", "filelist", "prefix=s", "nosort",
			 "quiet", "verbose", "v", "help", "debug", "trace")
	    || defined $opt_help ) {
	    &usage;
	}
	$opt_trace = 1 if defined $opt_debug;
	$opt_verbose = 0 if defined $opt_quiet;
	if ( defined $opt_prefix ) {
	    die ("$0: option \"-prefix\" requires \"-filelist\"\n")
		unless defined $opt_filelist;
	}
	if ( defined $opt_nosort ) {
	    die ("$0: option \"-nosort\" requires \"-filelist\"\n")
		unless defined $opt_filelist;
	}
	if ( defined $opt_filelist ) {
	    die ("$0: option \"-filelist\" only uses \"-manifest\"\n")
		if defined $opt_oldmanifest || defined $opt_newmanifest;
	}
	if ( defined $opt_manifest ) {
	    die ("$0: do not use \"-manifest\" with \"-oldmanifest\"".
		 " or \"-newmanifest\"\n")
		if defined $opt_newmanifest || defined $opt_oldmanifest;
	    $opt_newmanifest = $opt_oldmanifest = $opt_manifest;
	}
    }

    # Argument check.
    if ( defined $opt_filelist ) {
	if ( defined $opt_manifest ) {
	    &usage if @ARGV;
	    @ARGV = ( $opt_manifest );
	}
	else {
	    &usage unless @ARGV == 1;
	}
    }
    else {
	&usage unless @ARGV == 2;
    }
}

sub usage {
    print STDERR <<EoU;
This is $my_name version $my_version

Usage: $0 [options] old new
Usage: $0 -filelist [ -prefix XXX ] [ -nosort ] [ -manifest ] file

Makepatch options:
   -diff cmd		diff command to use, default \"$opt_diff\"
   -patchlevel file	file to use as patchlevel.h
   -man[ifest] file	list of files for old and new dir
   -newman[ifest] file	list of files for new dir
   -oldman[ifest] file	list of files for old dir
   -follow		follow symbolic links
Filelist options:
   -[file]list		extract filenames from manifest file
   -prefix XXX		add a prefix to these filenames
   -nosort		do not sort manifest entries
General options:
   -verbose		verbose output (default)
   -quiet		no verbose output
   -help		this message
EoU
    exit (1);
}
